### ISSUE TITLE
elasticsearch-maven-plugin and index lifecycle management support

### ISSUE THREAD

**Issue by annasarte** [alexcojocaru/elasticsearch-maven-plugin#99]
Hello.
We are running integration tests and are using the elasticsearch-maven-plugin.  Per the readme file, the "flavour" property defaults to "oss."  The "default" flavour is currently not working"due to xpack."  In our application, we are using ES life cycle management. And we are running into issues with the tests, in that they are failing with the error "unknown setting [index.lifecycle.name] please check that any required plugins are installed."  See stacktrace below.

Reading this blog https://discuss.elastic.co/t/unknown-setting-index-lifecycle-name-please-check-that-any-required-plugins/180091  it says **"oss" does not support index lifecycle management** and that is causing this error.

But since "oss" is the only supported "flavour" for the elasticsearch-maven-plugin, it seems we are out of luck.

Can you please advise if you are working on a fix for this, and also what we can do to solve this issue?  Here is a snippet of our integration test pom.xml:

`<plugins>
                    <plugin>
                        <groupId>com.github.alexcojocaru</groupId>
                        <artifactId>elasticsearch-maven-plugin</artifactId>
                        
                        <version>6.14</version>
                        <configuration>
                            
                            <version>6.7.1</version>
                            <clusterName>afm-integration-test</clusterName>
                            <transportPort>9301</transportPort>
                            <httpPort>9201</httpPort>
                            <setAwait>false</setAwait>
                        </configuration>`

Stacktrace:
15:31:49.590-04 [main] WARN  c.f.a.c.e.ElasticsearchIndexManager - RemoteTransportException[[JG12IGN][127.0.0.1:9301][indices:admin/template/put]]; nested: IllegalArgumentException[unknown setting [index.lifecycle.name] please check that any required plugins are installed, or check the breaking changes documentation for removed settings];
java.util.concurrent.ExecutionException: RemoteTransportException[[JG12IGN][127.0.0.1:9301][indices:admin/template/put]]; nested: IllegalArgumentException[unknown setting [index.lifecycle.name] please check that any required plugins are installed, or check the breaking changes documentation for removed settings];
	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:265)
	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:252)
	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:94)
	at com.fico.afm.common.elasticsearch.ElasticsearchIndexManager.createTemplates(ElasticsearchIndexManager.java:109)
	at com.fico.afm.common.elasticsearch.ElasticsearchIndexManager.create(ElasticsearchIndexManager.java:76)
	at com.fico.afm.common.service.AdminTenantBootstrapService.initializeAdminTenant(AdminTenantBootstrapService.java:39)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:363)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:307)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:136)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:414)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:307)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:303)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:110)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyPropertyValues(AbstractAutowireCapableBeanFactory.java:1681)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1433)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:592)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:303)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveValueIfNecessary(BeanDefinitionValueResolver.java:110)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyPropertyValues(AbstractAutowireCapableBeanFactory.java:1681)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1433)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:592)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.test.context.support.AbstractGenericContextLoader.loadContext(AbstractGenericContextLoader.java:128)
	at org.springframework.test.context.support.AbstractGenericContextLoader.loadContext(AbstractGenericContextLoader.java:60)
	at org.springframework.test.context.support.AbstractDelegatingSmartContextLoader.delegateLoading(AbstractDelegatingSmartContextLoader.java:275)
	at org.springframework.test.context.support.AbstractDelegatingSmartContextLoader.loadContext(AbstractDelegatingSmartContextLoader.java:243)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246)
	at org.springframework.test.context.testng.AbstractTestNGSpringContextTests.springTestContextPrepareTestInstance(AbstractTestNGSpringContextTests.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:564)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:213)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:138)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:175)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:107)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:281)
	at org.apache.maven.surefire.testng.TestNGXmlTestSuite.execute(TestNGXmlTestSuite.java:75)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:121)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)

Thank you in advance.
Anna

**Comment by alexcojocaru** [comment]
Using the plugin config in your comment, I cannot reproduce the issue; you either provided only part of the config and missed the key bits, either provided the wrong one.

Using a different config and setting the flavour to *default*, I was kind of able to get ES running - I had to set the exec permission on *target/elasticsearch0/platform/linux-x86_64/bin/controller* after unpacking the distro zip and before running the ES binary. The ES process started successfully, although I cannot say if the *default* specific functionality is really available or not - I presume not all, due to most of the binaries in the *modules* [sub]directories not having the exec flag set.

In summary, getting the *default* flavour to start can be easily done, but I am not confident that it would work fully. It looks like quite a challenge and, given how easily breaking changes are introduced by the ES devs even in minor releases (and not just major ones), maintaining that could also be quite a challenge.

If you are interested in building your own version of the plugin which would run the *default* flavour with the modules you need, let me know and I can help.

**Comment by annasarte** [comment]
Hello.  Yes here is some relevant code that uses life cycle management specifically.  
`private void putPolicy(final Tenant tenant, final AgingSetting setting) throws IOException {
        Map<String, Phase> phases = new HashMap<>();

        Map<String, LifecycleAction> rolloverActions = Collections.singletonMap(RolloverAction.NAME, new RolloverAction(null, rolloverMaxAge, null));
        phases.put("hot", new Phase("hot", null, rolloverActions));

        Map<String, LifecycleAction> deleteActions = Collections.singletonMap(DeleteAction.NAME, new DeleteAction());
        phases.put("delete", new Phase("delete", calculateDeleteAge(setting), deleteActions));

        LifecyclePolicy lifecyclePolicy = new LifecyclePolicy(policyName(tenant, setting), phases);
        PutLifecyclePolicyRequest putLifecyclePolicyRequest = new PutLifecyclePolicyRequest(lifecyclePolicy);
        restHighLevelClient.indexLifecycle().putLifecyclePolicy(putLifecyclePolicyRequest, RequestOptions.DEFAULT);

        Settings.Builder builder = Settings.builder()
                .put("index.lifecycle.name", policyName(tenant, setting))
                .put("index.lifecycle.rollover_alias", indexName(tenant, setting));
        UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest();
        updateSettingsRequest.indices(indexName(tenant, setting))
                .settings(builder);

        restHighLevelClient.indices().putSettings(updateSettingsRequest, RequestOptions.DEFAULT);
    }`

**Comment by alexcojocaru** [comment]
As stated in my previous comment, I am not willing to take on the challenge of having the plugin compatible with the default flavour at this point.
But, if you want to create a fork of this plugin which can start the default flavour for your use case, let me know and I can help.

**Comment by alexcojocaru** [comment]
I am closing this, since there hasn't any follow up in a long while.
